### DISCUSSION TITLE
ILI9341 Touchscreen Error

### DISCUSSION THREAD

**Discussion by MEnsar55** [micropython/micropython#11507] [RP2040 / Pico]
I'm using a Raspberry Pi Pico and an ili9341 touchscreen. To the codes below;

import machine
import ili9341

# Set up the SPI bus
spi = machine.SPI(0, baudrate=10000000, miso=machine.Pin(16), mosi=machine.Pin(19), 
sck=machine.Pin(18))

# Set up the LCD display
lcd = ili9341.ILI9341(spi, 240, 320, cs=machine.Pin(17), dc=machine.Pin(20))

# Clear the screen
lcd.clear()

# Write "Hello World" to the screen
lcd.write("Hello World")

# Display the screen
lcd.display()
i am getting this error: Traceback (most recent call last): File "", line 8, in AttributeError: 'module' object has no attribute 'ILI9341'

I downloaded the .py file in the link. https://github.com/rdagger/micropython-ili9341/blob/master/ili9341.py

**Comment by karfas** [comment]
>  am getting this error: Traceback (most recent call last): File "", line 8, in AttributeError: 'module' object has no attribute 'ILI9341'

What did you expect ? The file you linked doesn't define an `ILI9341` class; the only class in this file is named `Display` (and doesn't have a `write()` function).

**Comment by ctimmer** [comment]
I'm using this module on a pico w. Any of the demo's would have shown how to setup and use.

Also this [link](https://bytesnbits.co.uk/raspberry-pi-pico-spi-lcd-display/) was helpful.

Curt

**Reply by MEnsar55** [reply]
Well, I just want to print "Hello World" to the screen.  What should I do for this?

**Reply by robert-hh** [reply]
The driver you selected is for CircuitPython. It will not work with MicroPython,, unless it is the version modified for MicroPython. You can try the driver by @peterhinch in the drivers directory of https://github.com/peterhinch/micropython-nano-gui. I used that. It is shorter and Peter provides a good documentation.

**Reply by MEnsar55** [reply]
Well, what code did you run, can you share with me?

**Reply by robert-hh** [reply]
That's a litte piece of code showing some text. In addition to the ILI9341 driver, it uses the font file freeesans20.py and the library writer.py. All of that can be find in Peter's repositories. The font in gui/fonts, writer in https://github.com/peterhinch/micropython-font-to-py.
```
#
from machine import SPI, Pin
from ili9341 import ILI9341
from writer import CWriter
import freesans20
import time

backlight = Pin("LED_LCD", Pin.OUT)  # backlight is not a PWM pin
spi = SPI(7, sck="LCD_SCK", mosi="LCD_MOSI", miso="LCD_MISO", baudrate=24000000)
display = ILI9341(spi,  dc=Pin("LCD_D_C", Pin.OUT),
                  cs=Pin("LCD_CS", Pin.OUT, value=1),
                  rst=Pin("LCD_RESET", Pin.OUT, value=1))

BLACK = CWriter.create_color(display, 0, 0, 0, 0)
DARK_GREEN = CWriter.create_color(display, 5, 0, 128, 0)
GREEN = CWriter.create_color(display, 6, 0, 255, 0)
DARK_GRAY = CWriter.create_color(display, 12, 128, 128, 128)
YELLOW = CWriter.create_color(display, 14, 255, 255, 0)
WHITE = CWriter.create_color(display, 15, 255, 255, 255)

wri = CWriter(display, freesans20, fgcolor=WHITE, bgcolor=BLACK)
display.fill(0)
CWriter.set_textpos(display, 100, 60)  # verbose = False to suppress console output
wri.printstring("WIO Terminal D51R")
display.show()
backlight.on()
time.sleep(2)
```

**Reply by quantrpeter** [reply]
hi, i got this error

MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ImportError: can't import name ILI9341

**Reply by peterhinch** [reply]
The driver may be found [here](https://github.com/peterhinch/micropython-nano-gui/blob/master/drivers/ili93xx/ili9341.py) installed with
```bash
$ mpremote mip install "github:peterhinch/micropython-nano-gui/drivers/ili93xx"
```
However rather than trying to replicate the above code (which requires some further knowledge) I suggest approaching this via [nano-gui](https://github.com/peterhinch/micropython-nano-gui/tree/master): the docs indicate how to set up and use the display.

**Reply by quantrpeter** [reply]
thanks Peter

**Comment by robert-hh** [comment]
For the other I have a sample script as well. It needs a font driver and files as well: XglcdFont and the Unispace font file.
```
# MicroPython / Seeed Wio Terminal / SAMD51
# Wio-Terminal-Screen.py - output something on the ILI9341 screen
# scruss, 2022-10
# -*- coding: utf-8 -*-


from time import sleep
from ili9341_wio import Display, color565
from machine import Pin, SPI
from XglcdFont import XglcdFont

backlight = Pin("LED_LCD", Pin.OUT)  # backlight is not a PWM pin
spi = SPI(
    7, sck=Pin("LCD_SCK"), mosi=Pin("LCD_MOSI"), miso=Pin("LCD_MISO"), baudrate=24000000
)
display = Display(spi, dc=Pin("LCD_D/C"), cs=Pin("LCD_CS"), rst=Pin("LCD_RESET"),
                  rotation=90, width=320, height=240)
font = XglcdFont('Unispace12x24.c', 12, 24)
display.display_on()
display.clear()
backlight.on()

display.draw_text(50, 100, 'WIO Terminal D51R', font, color565(128, 128, 128))
```

**Reply by robert-hh** [reply]
Here's the full set for your initially used driver: 
[ili9341_pack.zip](https://github.com/micropython/micropython/files/11489862/ili9341_pack.zip)

**Reply by robert-hh** [reply]
You have in both cases to adapt the SPI pins.

**Reply by MEnsar55** [reply]
Ok, I'll try tomorrow.
